diff --git a/dist/nextjs/server/cookies.js b/dist/nextjs/server/cookies.js
index 8e3998668d674fe1a1dff3e834d5f561e9c8c5d3..cea5895625b98036b9095c3f1088567e6738bbd0 100644
--- a/dist/nextjs/server/cookies.js
+++ b/dist/nextjs/server/cookies.js
@@ -1,12 +1,12 @@
 import { cookies, headers } from "next/headers";
-export function getRequestCookies() {
-    return getCookieStore(headers(), cookies());
+export async function getRequestCookies() {
+    return getCookieStore(await headers(), await cookies());
 }
-export function getRequestCookiesInMiddleware(request) {
-    return getCookieStore(headers(), request.cookies);
+export async function getRequestCookiesInMiddleware(request) {
+    return getCookieStore(await headers(), request.cookies);
 }
-export function getResponseCookies(response) {
-    return getCookieStore(headers(), response.cookies);
+export async function getResponseCookies(response) {
+    return getCookieStore(await headers(), response.cookies);
 }
 function getCookieStore(requestHeaders, responseCookies) {
     const isLocalhost = /localhost:\d+/.test(requestHeaders.get("Host") ?? "");
diff --git a/dist/nextjs/server/index.js b/dist/nextjs/server/index.js
index 60dac150ae07c271f3ecf8b2ffcf49586cf286a9..899290d014eb767f2d2e5e6ac0e345b3051446d1 100644
--- a/dist/nextjs/server/index.js
+++ b/dist/nextjs/server/index.js
@@ -9,17 +9,18 @@ import { logVerbose, setAuthCookies, setAuthCookiesInMiddleware, } from "./utils
 /**
  * Wrap your app with this provider in your root `layout.tsx`.
  */
-export function ConvexAuthNextjsServerProvider(props) {
+export async function ConvexAuthNextjsServerProvider(props) {
     const { apiRoute, storage, storageNamespace, verbose, children } = props;
-    return (_jsx(ConvexAuthNextjsClientProvider, { serverState: convexAuthNextjsServerState(), apiRoute: apiRoute, storage: storage, storageNamespace: storageNamespace, verbose: verbose, children: children }));
+    const serverState = await convexAuthNextjsServerState();
+    return (_jsx(ConvexAuthNextjsClientProvider, { serverState: serverState, apiRoute: apiRoute, storage: storage, storageNamespace: storageNamespace, verbose: verbose, children: children }));
 }
 /**
  * Retrieve the token for authenticating calls to your
  * Convex backend from Server Components, Server Actions and Route Handlers.
  * @returns The token if the the client is authenticated, otherwise `undefined`.
  */
-export function convexAuthNextjsToken() {
-    return getRequestCookies().token ?? undefined;
+export async function convexAuthNextjsToken() {
+    return (await getRequestCookies()).token ?? undefined;
 }
 /**
  * Whether the client is authenticated, which you can check
@@ -28,8 +29,8 @@ export function convexAuthNextjsToken() {
  * Avoid the pitfall of checking authentication state in layouts,
  * since they won't stop nested pages from rendering.
  */
-export function isAuthenticatedNextjs() {
-    return convexAuthNextjsToken() !== undefined;
+export async function isAuthenticatedNextjs() {
+    return (await convexAuthNextjsToken()) !== undefined;
 }
 /**
  * Use in your `middleware.ts` to enable your Next.js app to use
@@ -38,11 +39,11 @@ export function isAuthenticatedNextjs() {
  * @returns A Next.js middleware.
  */
 export function convexAuthNextjsMiddleware(
-/**
- * A custom handler, which you can use to decide
- * which routes should be accessible based on the client's authentication.
- */
-handler, options = {}) {
+    /**
+     * A custom handler, which you can use to decide
+     * which routes should be accessible based on the client's authentication.
+     */
+    handler, options = {}) {
     return async (request, event) => {
         const verbose = options.verbose ?? false;
         logVerbose(`Begin middleware for request with URL ${request.url}`, verbose);
@@ -66,7 +67,7 @@ handler, options = {}) {
         if (authResult.kind === "refreshTokens" &&
             authResult.refreshTokens !== undefined) {
             logVerbose(`Forwarding cookies to request`, verbose);
-            setAuthCookiesInMiddleware(request, authResult.refreshTokens);
+            await setAuthCookiesInMiddleware(request, authResult.refreshTokens);
         }
         if (handler === undefined) {
             logVerbose(`No custom handler`, verbose);
@@ -81,25 +82,25 @@ handler, options = {}) {
                 (await handler(request, {
                     event,
                     convexAuth: {
-                        getToken: () => {
-                            const cookies = getRequestCookiesInMiddleware(request);
+                        getToken: async () => {
+                            const cookies = await getRequestCookiesInMiddleware(request);
                             return cookies.token ?? undefined;
                         },
-                        isAuthenticated: () => {
-                            const cookies = getRequestCookiesInMiddleware(request);
+                        isAuthenticated: async () => {
+                            const cookies = await getRequestCookiesInMiddleware(request);
                             return (cookies.token ?? undefined) !== undefined;
                         },
                     },
                 })) ??
-                    NextResponse.next({
-                        headers: request.headers,
-                    });
+                NextResponse.next({
+                    headers: request.headers,
+                });
         }
         // Port the cookies from the auth middleware to the response
         if (authResult.kind === "refreshTokens" &&
             authResult.refreshTokens !== undefined) {
             const nextResponse = NextResponse.next(response);
-            setAuthCookies(nextResponse, authResult.refreshTokens);
+            await setAuthCookies(nextResponse, authResult.refreshTokens);
             return nextResponse;
         }
         return response;
@@ -115,20 +116,20 @@ export { createRouteMatcher } from "./routeMatcher.js";
  * ```
  */
 export function nextjsMiddlewareRedirect(
-/**
- * The incoming request handled by the middleware.
- */
-request, 
-/**
- * The route path to redirect to.
- */
-pathname) {
+    /**
+     * The incoming request handled by the middleware.
+     */
+    request,
+    /**
+     * The route path to redirect to.
+     */
+    pathname) {
     const url = request.nextUrl.clone();
     url.pathname = pathname;
     return NextResponse.redirect(url);
 }
-function convexAuthNextjsServerState() {
-    const { token } = getRequestCookies();
+async function convexAuthNextjsServerState() {
+    const { token } = await getRequestCookies();
     return {
         // The server doesn't share the refresh token with the client
         // for added security - the client has to use the server
diff --git a/dist/nextjs/server/proxy.js b/dist/nextjs/server/proxy.js
index a53422955dd90fec325eef232fb4eb4844b5d071..61c3a82aa19969a1d24c1992413dc227c2e2f931 100644
--- a/dist/nextjs/server/proxy.js
+++ b/dist/nextjs/server/proxy.js
@@ -19,7 +19,7 @@ export async function proxyAuthActionToConvex(request, options) {
     if (action === "auth:signIn" && args.refreshToken !== undefined) {
         // The client has a dummy refreshToken, the real one is only
         // stored in cookies.
-        const refreshToken = getRequestCookies().refreshToken;
+        const refreshToken = (await getRequestCookies()).refreshToken;
         if (refreshToken === null) {
             console.error("Convex Auth: Unexpected missing refreshToken cookie during client refresh");
             return new Response(JSON.stringify({ tokens: null }));
@@ -30,7 +30,7 @@ export async function proxyAuthActionToConvex(request, options) {
         // Make sure the proxy is authenticated if the client is,
         // important for signOut and any other logic working
         // with existing sessions.
-        token = getRequestCookies().token ?? undefined;
+        token = (await getRequestCookies()).token ?? undefined;
     }
     logVerbose(`Fetching action ${action} with args ${JSON.stringify(args)}`, verbose);
     if (action === "auth:signIn") {
@@ -51,13 +51,13 @@ export async function proxyAuthActionToConvex(request, options) {
             console.error(error);
             logVerbose(`Clearing auth cookies`, verbose);
             const response = jsonResponse(null);
-            setAuthCookies(response, null);
+            await setAuthCookies(response, null);
             return response;
         }
         if (result.redirect !== undefined) {
             const { redirect } = result;
             const response = jsonResponse({ redirect });
-            getResponseCookies(response).verifier = result.verifier;
+            (await getResponseCookies(response)).verifier = result.verifier;
             logVerbose(`Redirecting to ${redirect}`, verbose);
             return response;
         }
@@ -73,7 +73,7 @@ export async function proxyAuthActionToConvex(request, options) {
                     ? { token: result.tokens.token, refreshToken: "dummy" }
                     : null,
             });
-            setAuthCookies(response, result.tokens);
+            await setAuthCookies(response, result.tokens);
             return response;
         }
         return jsonResponse(result);
@@ -91,7 +91,7 @@ export async function proxyAuthActionToConvex(request, options) {
         }
         logVerbose(`Clearing auth cookies`, verbose);
         const response = jsonResponse(null);
-        setAuthCookies(response, null);
+        await setAuthCookies(response, null);
         return response;
     }
 }
diff --git a/dist/nextjs/server/request.js b/dist/nextjs/server/request.js
index 88ec2f68ab4a17ab4bc7cf48d997b39ad47bfeef..cb6b246a60dc2d8e427a7d922238e8355b26df32 100644
--- a/dist/nextjs/server/request.js
+++ b/dist/nextjs/server/request.js
@@ -7,7 +7,7 @@ export async function handleAuthenticationInRequest(request, verbose) {
     logVerbose(`Begin handleAuthenticationInRequest`, verbose);
     const requestUrl = new URL(request.url);
     // Validate CORS
-    validateCors(request);
+    await validateCors(request);
     // Refresh tokens if necessary
     const refreshTokens = await getRefreshedTokens(verbose);
     // Handle code exchange for OAuth and magic links via server-side redirect
@@ -16,7 +16,7 @@ export async function handleAuthenticationInRequest(request, verbose) {
         request.method === "GET" &&
         request.headers.get("accept")?.includes("text/html")) {
         logVerbose(`Handling code exchange for OAuth or magic link`, verbose);
-        const verifier = getRequestCookies().verifier ?? undefined;
+        const verifier = (await getRequestCookies()).verifier ?? undefined;
         const redirectUrl = new URL(requestUrl);
         redirectUrl.searchParams.delete("code");
         try {
@@ -25,7 +25,7 @@ export async function handleAuthenticationInRequest(request, verbose) {
                 throw new Error("Invalid `signIn` action result for code exchange");
             }
             const response = NextResponse.redirect(redirectUrl);
-            setAuthCookies(response, result.tokens);
+            await setAuthCookies(response, result.tokens);
             logVerbose(`Successfully validated code, redirecting to ${redirectUrl.toString()} with auth cookies`, verbose);
             return { kind: "redirect", response };
         }
@@ -33,7 +33,7 @@ export async function handleAuthenticationInRequest(request, verbose) {
             console.error(error);
             logVerbose(`Error validating code, redirecting to ${redirectUrl.toString()} and clearing auth cookies`, verbose);
             const response = NextResponse.redirect(redirectUrl);
-            setAuthCookies(response, null);
+            await setAuthCookies(response, null);
             return { kind: "redirect", response };
         }
     }
@@ -41,9 +41,9 @@ export async function handleAuthenticationInRequest(request, verbose) {
 }
 // If this is a cross-origin request with `Origin` header set
 // do not allow the app to read auth cookies.
-function validateCors(request) {
+async function validateCors(request) {
     if (isCorsRequest(request)) {
-        const cookies = getRequestCookiesInMiddleware(request);
+        const cookies = await getRequestCookiesInMiddleware(request);
         cookies.token = null;
         cookies.refreshToken = null;
         cookies.verifier = null;
@@ -52,7 +52,7 @@ function validateCors(request) {
 const REQUIRED_TOKEN_LIFETIME_MS = 60_000; // 1 minute
 const MINIMUM_REQUIRED_TOKEN_LIFETIME_MS = 10_000; // 10 seconds
 async function getRefreshedTokens(verbose) {
-    const cookies = getRequestCookies();
+    const cookies = await getRequestCookies();
     const { token, refreshToken } = cookies;
     if (refreshToken === null && token === null) {
         logVerbose(`No tokens to refresh, returning undefined`, verbose);
diff --git a/dist/nextjs/server/utils.js b/dist/nextjs/server/utils.js
index eecdc5b3dbb6b7f6acdb2d5e546e82e99a42d30c..2697fc6680fbcdc9473e6c037d07ccc52f99ebbd 100644
--- a/dist/nextjs/server/utils.js
+++ b/dist/nextjs/server/utils.js
@@ -5,8 +5,8 @@ export function jsonResponse(body) {
         headers: { "Content-Type": "application/json" },
     });
 }
-export function setAuthCookies(response, tokens) {
-    const responseCookies = getResponseCookies(response);
+export async function setAuthCookies(response, tokens) {
+    const responseCookies = await getResponseCookies(response);
     if (tokens === null) {
         responseCookies.token = null;
         responseCookies.refreshToken = null;
@@ -23,8 +23,8 @@ export function setAuthCookies(response, tokens) {
  * @param request
  * @param tokens
  */
-export function setAuthCookiesInMiddleware(request, tokens) {
-    const requestCookies = getRequestCookiesInMiddleware(request);
+export async function setAuthCookiesInMiddleware(request, tokens) {
+    const requestCookies = await getRequestCookiesInMiddleware(request);
     if (tokens === null) {
         requestCookies.token = null;
         requestCookies.refreshToken = null;
